talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,166 @@
-          if (errno) {
-            throw string("Bad Value");
-          }
-          divide(100, l);
-        } catch (string& e) {
-          printf("%s\n", e.c_str());
-        } catch (ZeroDivide e2) {
-          printf("Division error\n");
+    \begin{cppcode*}{fontsize=\scriptsize,gobble=2}
+      int process_file(const fileobj file) {
+        if (fh = open_file(file)) {
+          throw
+            file_bad(file.error.reason());
+        }
+        while (!fh) {
+          line = read_line(fh);
+          add_line_to_db(line); // Can throw
+                                // db_bad
         }
+        return 0;
       }
     \end{cppcode*}
   \end{multicols}
 \end{frame}
 
-\begin{frame}[fragile]
-  \frametitlecpp[98]{Controlling exceptions \hfill \deprecated/\removed}
-  \begin{block}{Declaring expected exceptions}
-    \begin{itemize}
-    \item each function can declare a set of expected exceptions
-    \item using the {\it throw} statement in its declaration
-    \item other exceptions won't exit the scope of the function
-      \begin{itemize}
-      \item instead, the {\it unexpected} handler is called
-      \item by default, it terminates the program
-      \end{itemize}
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{cppcode}
-    int func(int a) throw(int) {
-      if (0 == a) {
-        throw 2;  // ok, goes out
-      } else {
-        throw "hello"; // std::unexpected called
-      }
-    }
-  \end{cppcode}
-\end{frame}
-
-\begin{frame}[fragile]
-  \frametitlecpp[98]{Controlling exceptions \hfill \deprecated/\removed}
-  \begin{block}{Good to know}
-    \begin{itemize}
-    \item The check is done at runtime, not at compile time
-      \begin{itemize}
-      \item unlike Java
-      \end{itemize}
-    \item When the {\it throw} clause is absent, any exception can go out
-    \item To block all exceptions, use {\it throw()}
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{cppcode}
-    int func(int a) {
-      // any exception can go out
-    }
-    int otherfunc(int a) throw() {
-      // no exception can go out
-    }
-  \end{cppcode}
-\end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[11]{Removal of \cpp98 exceptions}
+  \frametitlecpp[98]{Error Handling and Exceptions}
   \begin{block}{}
-    After a lot of thinking and experiencing, the conclusions of the community on exception handling are :
     \begin{itemize}
-    \item Never write an exception specification
-    \item Except possibly an empty one
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{alertblock}{Some of the reasons}
-    \begin{itemize}
-    \item throw specification is runtime only
+      \item exceptions have little cost if no exception is thrown
       \begin{itemize}
-      \item does not allow compiler optimizations
-      \item on the contrary forces extra checks
-      \item generally terminates your program if violated
+        \item a throw usually costs about as much as a function call
```

Comment:
That depends on a lot of factors. I cannot find the paper on performance of various kind of error handling, but I remember exceptions where crazy expensive if you throw from a deeper call hierarchy or if you throw polymorphic values. But I cannot request you to change your wording until I find the numbers :/

Reply:
Throw might require you to unwind of stuff and tear it down properly. The statement about the low cost when they are rare likely holds, though.
Just don't abuse them to pass information around.

Reply:
Based on : https://docs.microsoft.com/en-us/cpp/cpp/errors-and-exception-handling-modern-cpp?view=msvc-170#exceptions-and-performance - could be compiler specific of course. I found a few stack overflow articles claiming the same, that the cost is throwing the exception, not the try ... catch. Throwing unwinds the stack, but that would be just as expensive when passed up through error values as done with an exception (and then also far more fragile).

Reply:
Rephrased this to be a bit clearer that the throw is ~like an *extra* call and that the stack always needs to be unwound